models/search.js

```diff
@@ -0,0 +1,130 @@
+  }
+}
+
+function generateOrderByClause(sortBy, tsQuery) {
+  switch (sortBy) {
+    case 'new':
+      return 'ORDER BY created_at DESC';
+    case 'relevant':
+      return `ORDER BY ts_rank(ts, ${tsQuery}) DESC`;
+    case 'old':
+      return 'ORDER BY created_at ASC';
+    default:
+      return '';
+  }
+}
+
+async function findAll(values = {}) {
+  const query = {
+    values: [],
+  };
+
+  if (!values.count) {
+    const offset = (values.page - 1) * values.per_page;
+    query.values = [values.limit || values.per_page, offset];
+  }
+
+  const separatedWords = generateSearchTerm(values.q);
+  const tsQuery = generateTsQuery(separatedWords);
+  const selectClause = buildSelectClause(values, tsQuery);
+
+  query.text = `
+    WITH content_window AS (
+      SELECT
+        COUNT(*) OVER()::INTEGER as total_rows,
+        id
+      FROM 
+        contents
+      WHERE
+        to_tsvector('portuguese', contents.title) @@ ${tsQuery}
+      AND 
+        contents.status = 'published'
+      ${generateOrderByClause(values.sort, tsQuery)}
+      ${values.count ? 'LIMIT 1' : 'LIMIT $1 OFFSET $2'}
+    )
+    ${selectClause}
+  `;
+
+  function buildSelectClause(values, tsQuery) {
+    if (values.count) {
+      return `
+        SELECT
+          total_rows
+        FROM
+          content_window
+        `;
+    }
+
+    return `
+      SELECT
+        contents.id,
+        contents.owner_id,
+        contents.parent_id,
+        contents.slug,
+        contents.title,
+        contents.status,
+        contents.source_url,
+        contents.created_at,
+        contents.updated_at,
+        contents.published_at,
+        contents.deleted_at,
+        contents.path,
+        users.username as owner_username,
+        content_window.total_rows,
+        tabcoins_count.total_balance as tabcoins,
+        tabcoins_count.total_credit as tabcoins_credit,
+        tabcoins_count.total_debit as tabcoins_debit,
+        (
+          SELECT COUNT(*)
+          FROM contents as children
+          WHERE children.path @> ARRAY[contents.id]
+           AND children.status = 'published'
+        ) as children_deep_count
+      FROM
+        contents
+      INNER JOIN
+        content_window ON contents.id = content_window.id
+      INNER JOIN
+        users ON contents.owner_id = users.id
+      LEFT JOIN LATERAL 
+        get_content_balance_credit_debit(contents.id) tabcoins_count ON true
+      WHERE 
+        contents.status = 'published'
+      AND 
+        to_tsvector('portuguese', contents.title) @@ ${tsQuery}
+      ${generateOrderByClause(values.sort, tsQuery)}
+    `;
+  }
+
+  const queryResults = await database.query(query);
+
+  const results = {
+    rows: queryResults.rows,
+  };
+
+  values.total_rows = results.rows[0]?.total_rows;
+  results.pagination = await getPagination(values);
+
+  return results;
+}
+
+async function getPagination(values) {
+  values.count = true;
+  return pagination.get(values);
+}
```

Comment:
2 detalhes aqui:

1. A atribuição de `values.count` não faz diferença para `pagination.get`.
2. `pagination.get` não é assíncrono.

```suggestion
  results.pagination = await pagination.get(values);

  return results;
}
```

Reply:
pagination.get não é assíncrono ou não assíncrono?

Reply:
Não é, corrigi meu comentário.

Reply:
Se values.count não faz diferença, onde tem **values.count** vai ser sempre null?

![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/ebafab84-cde1-48d4-9673-c841a9cad7a4)

![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/656f705c-a9b5-44df-b1ac-f46dc37a5093)

![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/01268031-edb9-498d-a7e0-c212de8b0da5)

Reply:
@Rafatcb sem passar await no inicio dele, ele falha, retorna nada.

com: 
![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/65c651cf-0cb4-4fb1-b8a6-cdd149ceca0b)

sem:
![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/d740f01f-9484-47e2-b4df-90b52d2072d5)

Reply:
ele é async, talvez precisa ser mudado:

![image](https://github.com/filipedeschamps/tabnews.com.br/assets/121470728/61257edf-8c49-490f-8ae3-7b57ef41f07a)

Reply:
Obrigado por perceber isso. Já abri um PR para corrigir https://github.com/filipedeschamps/tabnews.com.br/pull/1707, então você conseguirá chamar a função sem `await`.

Reply:
O PR #1707 foi mesclado, quando você fizer o `rebase` já conseguirá tirar o `await` aqui.